reservoir-backend.js

```diff
@@ -181,26 +182,68 @@ ReservoirBackend.prototype.flush = function flush() {
     self.count = 0;
 };
 
+ReservoirBackend.prototype.willSample = function willSample(level) {
+    this.samplingDecision = this._makeSamplingDecision(level);
+    if (this.samplingDecision !== -1) {
+        return true;
+    } else {
+        return false;
+    }
+};
+
+ReservoirBackend.prototype._makeSamplingDecision =
+function _makeSamplingDecision(level) {
+    if (this.records.length < this.size) {
+        return this.records.length;
+    }
+
+    var probability = this.rangeRand(0, this.count);
+    if (probability < this.size) {
+        return probability;
+    } else {
+        return -1;
+    }
+};
+
 ReservoirBackend.prototype.log = function log(record, cb) {
-    var self = this;
+    // If we don't already have a sampling decision, make one
+    var decision = this._makeSamplingDecision(record.data.level);
 
-    self.count += 1;
+    this.count += 1;
 
-    if (self.records.length < self.size) {
-        self.records.push(record);
+    if (decision !== -1) {
+        if (this.records[decision]) {
```

Comment:
micro-op: do a length check instead ?

Accessing an index out of bounds will de-op right ?